example/linux/gw-example/main.c

```diff
@@ -155,9 +195,20 @@ static void * publish_thread(void * unused)
 
 static void on_mqtt_message_delivered(void *context, MQTTClient_deliveryToken dt)
 {
-    // For now, only informal but message should be cleared from queue only when
-    // delivered and not published as of today
+    // Warning : this is called only if Qos is not 0
     LOGI("Message with token %d delivery confirmed\n", dt);
+    for (unsigned int i = 0; i < PUBLISH_QUEUE_SIZE; i++)
+    {
+        if (m_publish_queue[i].token == dt)
+        {
+            // Free the payload
+            free(m_publish_queue[i].payload_p);
```

Comment:
I don't see the point to free it there if you don't have a way to republish or insert in the queue at random place.
But as it is not supported yet, you should remove it directly once publish is done.

Reply:
But now I understand better why you check if payload is not null :-)

Reply:
And it seems that it could create issue.
Let's imagine you receive the notification for a message that you already replace the payload but not publish yet so token is still the old one, then new payload will be cleared.